### PR TITLE
Gateway testnet

### PR DESCRIPTION
**Pull request by fadeev**
At the moment gateway is only enabled on Base Sepolia, Polygon Amoy and ZetaChain testnet.

Gateway on ZetaChain: `0x6c533f7fe93fae114d0954697069df33c9b74fd7`
Gateway on EVM chains: `0x0c487a766110c85d301d96e33579c5b317fa4995`

## Hello Example

```
cd examples/hello
yarn
yarn deploy
```

```
🔑 Using account: 0x4955a3F38ff86ae92A914445099caa8eA2B9bA32

🚀 Successfully deployed "Hello" contract on zeta_testnet.
📜 Contract address: 0x496CD66950a1F1c5B02513809A2d37fFB942be1B

🔑 Using account: 0x4955a3F38ff86ae92A914445099caa8eA2B9bA32

🚀 Successfully deployed "Echo" contract on base_sepolia.
📜 Contract address: 0x775329c70A8d09AEb5e5ca92C369FF3155C5f1Ed
```

### Call from Base to ZetaChain

```
npx hardhat echo-call \
  --contract 0x775329c70A8d09AEb5e5ca92C369FF3155C5f1Ed \
  --receiver 0x496CD66950a1F1c5B02513809A2d37fFB942be1B \
  --network base_sepolia \
  --tx-options-gas-price 20000 --types '["string"]' alice
```

https://sepolia.basescan.org/tx/0x133cdf3a06195de0a6bb89dd4761ca98d1301534b3c4987f0ff93c95c3fff78c

✅ https://zetachain-athens.blockpi.network/lcd/v1/public/zeta-chain/crosschain/inboundHashToCctxData/0x133cdf3a06195de0a6bb89dd4761ca98d1301534b3c4987f0ff93c95c3fff78c

### Call from ZetaChain to Base

Had to specify gas price — need to update the defaults to match testnet:

```
npx hardhat hello-call \
  --contract 0x496CD66950a1F1c5B02513809A2d37fFB942be1B \
  --receiver 0x775329c70A8d09AEb5e5ca92C369FF3155C5f1Ed \
  --zrc20 0x236b0DE675cC8F46AE186897fCCeFe3370C9eDeD \
  --function "hello(string)" \
  --network zeta_testnet \
  --tx-options-gas-price 200000 --types '["string"]' alice
```

https://zetachain-testnet.blockscout.com/tx/0x19d476fa2c3d29ba41467ae7f2742541fd11e0b67d6548fe7655a3d40274323e

✅ https://zetachain-athens.blockpi.network/lcd/v1/public/zeta-chain/crosschain/inboundHashToCctxData/0x19d476fa2c3d29ba41467ae7f2742541fd11e0b67d6548fe7655a3d40274323e

### Withdraw And Call from ZetaChain to Base

```
npx hardhat hello-withdraw-and-call \
  --contract 0x496CD66950a1F1c5B02513809A2d37fFB942be1B \
  --receiver 0x775329c70A8d09AEb5e5ca92C369FF3155C5f1Ed \
  --zrc20 0x236b0DE675cC8F46AE186897fCCeFe3370C9eDeD \
  --function "hello(string)" \
  --amount 0.005 \
  --network zeta_testnet \
  --types '["string"]' hello
```

https://zetachain-testnet.blockscout.com/tx/0x67099389ab6cb44ee03602d164320c615720b57762c5ddab042d65bdbe30c7a2

✅ https://zetachain-athens.blockpi.network/lcd/v1/public/zeta-chain/crosschain/inboundHashToCctxData/0x67099389ab6cb44ee03602d164320c615720b57762c5ddab042d65bdbe30c7a2

### Deposit from Base to ZetaChain

Not related to examples.

```
npx hardhat evm-deposit --amount 0.05 --gateway-evm 0x0c487a766110c85d301d96e33579c5b317fa4995 --receiver 0x4955a3F38ff86ae92A914445099caa8eA2B9bA32 --network base_sepolia --gas-price 200000

Transaction hash: 0xbb864575f2c4340e868946649c7a25f954b1d3804aed4f5b11bef09cb3d75ca4
```

https://sepolia.basescan.org/tx/0xbb864575f2c4340e868946649c7a25f954b1d3804aed4f5b11bef09cb3d75ca4

✅ https://zetachain-athens.blockpi.network/lcd/v1/public/zeta-chain/crosschain/inboundHashToCctxData/0xbb864575f2c4340e868946649c7a25f954b1d3804aed4f5b11bef09cb3d75ca4

## Swap

```
yarn deploy
```

```
🔑 Using account: 0x4955a3F38ff86ae92A914445099caa8eA2B9bA32

🚀 Successfully deployed contract on zeta_testnet.
📜 Contract address: 0x162CefCe314726698ac1Ee5895a6c392ba8e20d3
```

### Deposit and Call from Base to ZetaChain to Amoy

```
npx hardhat evm-deposit-and-call --receiver 0x162CefCe314726698ac1Ee5895a6c392ba8e20d3 --amount 0.001 --types '["address", "bytes"]' 0x777915D031d1e8144c90D025C594b3b8Bf07a08d 0x4955a3F38ff86ae92A914445099caa8eA2B9bA32 --network base_sepolia --gas-price 20000 --gateway-evm 0x0c487a766110c85d301d96e33579c5b317fa4995
```

https://sepolia.basescan.org/tx/0xc81cc6d1383fb55bb99d0c672af1ab1db2dce003f3dc00a32e316e469733530c

https://zetachain-athens.blockpi.network/lcd/v1/public/zeta-chain/crosschain/inboundHashToCctxData/0xc81cc6d1383fb55bb99d0c672af1ab1db2dce003f3dc00a32e316e469733530c

Added NFT example from https://github.com/zeta-chain/example-contracts/pull/206

Depends on https://github.com/zeta-chain/example-contracts/pull/207

## Release Notes

- **New Features**
	- Introduced new NFT and swap contracts with enhanced cross-chain capabilities.
	- Added tasks for deploying, minting, transferring NFTs, and setting counterparty addresses.
	- Enhanced deployment scripts for local networks.
	- Added a new `.eslintignore` file to streamline linting processes.
	- Introduced a new `foundry.toml` configuration file for Foundry framework settings.
	- New Bash scripts for testing and deploying contracts on local networks.
	- Added a new section in README files to clarify compatibility with localnet version `v4.0.0-rc*`.

- **Bug Fixes**
	- Updated function signatures for improved context handling in cross-chain operations.

- **Documentation**
	- Added new sections in README files to clarify compatibility and usage.
	- Introduced new configuration files for TypeScript and ESLint to improve development experience.
	- Added a new MIT License file to the NFT example project.

- **Chores**
	- Updated dependency versions for improved stability and functionality across the project.
	- Added a new GitHub Actions workflow for automated testing.

### CODE REVIEW

**Comment by fadeev** [comment]
I'm having issues with verification, so contracts are not yet verified.

**Comment by fadeev** [comment]
@andresaiello please, review.

**Review by andresaiello — APPROVED** [review]

File: .github/workflows/test.yaml [hunk 1]
```diff
@@ -0,0 +1,43 @@
+name: Run Tests
+
+on:
+  push:
+  workflow_dispatch:
+
```
[the remark this answers is not in the document]
Comment (gartnera) [reply .github/workflows/test.yaml:6]: any reason to not include `pull_request`?

File: .github/workflows/test.yaml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v3
+
+      - name: Set up Test Matrix
+        id: set-matrix
+        run: |
+          test_dirs=$(find examples/*/scripts -type f -name 'test.sh' -exec dirname {} \; | xargs dirname)
+          matrix_json=$(echo "$test_dirs" | jq -R '{"example-dir": .}' | jq -s . | jq -c .)
+          echo "matrix=$matrix_json" >> $GITHUB_OUTPUT
```
Comment (gartnera) [inline .github/workflows/test.yaml:21]: might be cleaner to just write this in actions/github-script in javascript: https://github.com/actions/github-script

File: .github/workflows/test.yaml
Comment (gartnera) [inline .github/workflows/test.yaml:1]: > The following actions use a deprecated Node.js version and will be forced to run on node20: actions/checkout@v3

Comment (fadeev) [reply .github/workflows/test.yaml:1]: https://github.com/zeta-chain/example-contracts/pull/207/commits/1ed0d9e183eccce071a04f40f74e391e9fc6e281

**Pull request merged**